Hilbert curve followup; G26 change for additional testing

marciot opened the pull request:
- Fixed probe offset being applied twice, which caused "G29 P1 X0 Y0" to start in the wrong location
   - The offset was being applied on line https://github.com/MarlinFirmware/Marlin/blob/1caf8a1f5b1aaaca01e10af2a918a264ed549a1b/Marlin/src/feature/bedlevel/ubl/ubl_G29.cpp#L436 and then again on line  https://github.com/MarlinFirmware/Marlin/blob/1caf8a1f5b1aaaca01e10af2a918a264ed549a1b/Marlin/src/feature/bedlevel/ubl/ubl_G29.cpp#L1324
- Changed the default position for a bed probe with UBL_HILBERT_CURVE to X0 Y0
- Allow G26 to use a Hilbert curve&Dagger;

&Dagger; The change to G26 had been included in #21453, but @GhostlyCrowd indicated the change had broken G26, so it was removed from that PR.

This PR re-introduces those changes for additional testing and verification. I tested out "G26 R" and found that it behaved as expected. Plain "G26" requires a different type of display than I have, so I am hoping someone else can test that.

[Concise Diff](21480/files?w=1)

marciot committed:
Fix for "G29 P1 X0 Y0" not starting in correct grid point

- Fixed probe offset being applied twice

marciot committed:
Start UBL_HILBERT_CURVE at lower-left

marciot committed:
Allow G26 to use UBL_HILBERT_CURVE

GhostlyCrowd commented:
Just testing now, still not doing a validation mesh with `G26 P10`

marciot commented:
> Just testing now, still not doing a validation mesh with `G26 P10`

Does it work when UBL_HILBERT_CURVE is disabled? Also, what does P10 do?

GhostlyCrowd commented:
> > Just testing now, still not doing a validation mesh with `G26 P10`
> 
> Does it work when UBL_HILBERT_CURVE is disabled? Also, what does P10 do?

I will try with it disabled, and P10 purges 10mm of material to prime the nozzle before the mesh is printed.

GhostlyCrowd commented:
`G26 P10` does with with `UBL_HILBERT_CURVE` disabled.

thinkyhead committed:
reformat

thinkyhead committed:
Use less SRAM

thinkyhead committed:
etc.

thinkyhead committed:
Merge remote-tracking branch 'upstream/bugfix-2.0.x' into pr/21480

thinkyhead committed:
Merge remote-tracking branch 'upstream/bugfix-2.0.x' into pr/21480

thinkyhead committed:
Put G26 in a struct

thinkyhead committed:
prettier

marciot commented:
@GhostlyCrowd: I'm still stumped by this. It seems to work for me and I can see no reason why it would not work for you. Can you try the following things:

1) Does it work for your with the "R" flag?
2) Does it work if you move the nozzle to different positions before starting "G26"?
3) Does it work if you use the "X" and "Y" parameters to select different start positions?

marciot committed:
Fix compilation error.

GhostlyCrowd commented:
> @GhostlyCrowd: I'm still stumped by this. It seems to work for me and I can see no reason why it would not work for you. Can you try the following things:
> 
> 1. Does it work for your with the "R" flag?
> 2. Does it work if you move the nozzle to different positions before starting "G26"?
> 3. Does it work if you use the "X" and "Y" parameters to select different start positions?

I'll give this a try tonight. and let you know. 
I'm currently in my office.

thinkyhead committed:
move some things

thinkyhead committed:
missing prefix

GhostlyCrowd commented:
`Marlin/src/gcode/bedlevel/G26.cpp:472:7: error: expected ',' or ';' before 'hilbert_curve'` I put in a ; and it compiled no idea if thats correct. Testing now. But im unsure if it will work.

GhostlyCrowd commented:
Ok so G26 P1 now works with the small code fix (adding ;) but its printing "extra" what i mean by extra is stuff that shouldn't be part of the mesh validation and is out of bounds of the mesh or the margins i set.  I have a 30mm margin on all 4 sides of the mesh and its running the "extra" mesh off Y max.  I will edit this with a picture when its done.  And because its printing phantom mesh its actually breaking the efficiency of the hilbert G26, until this phantom meshing its printing efficiently and the moment is more efficient then current G26

GhostlyCrowd commented:
More findings and probably why i couldn't get `G26` to work before and you could. `G26` Fails if there is ANY movement or other commands sent to the printer, the only way i can get `G26 P10` to work is if i home exclusivity directly before it, `G28` then `G26 P10`. 

Macro i use is below
`G28
G34
G29 P1
G29 P3
G29 S1
G29 F10
G29 A
M500
G29 L1
G1 X0 Y0 F60000
G26 P10 
M400
G29 S1`
So you can understand why my macro to probe and validate didn't work. The original G26 code didnt behave this way.

thinkyhead committed:
add ;

thinkyhead commented:
Given how "inexpensive" the Hilbert Curve is, compared to UBL and/or `G26` as a whole, maybe it makes sense to make it an option (`G26 H1`) while still offering the standard `G26` as well. The default could be `H1`, requiring `H0` to *not* use the Hilbert Curve, in that case.

marciot commented:
> Given how "inexpensive" the Hilbert Curve is, compared to UBL and/or `G26` as a whole, maybe it makes sense to make it an option (`G26 H1`) while still offering the standard `G26` as well. The default could be `H1`, requiring `H0` to _not_ use the Hilbert Curve, in that case.

It’s not particularly inexpensive in terms of stack space. It’s a recursive function with about eight parameters, so I might push the limit on small AVR chips. So I don’t recommend making it default, but it could be made an option.

I can’t think of a reason why having both available at runtime is an advantage though.

thinkyhead committed:
finish up

marciot commented:
> Ok so G26 P1 now works with the small code fix (adding ;) but its printing "extra" what i mean by extra is stuff that shouldn't be part of the mesh validation and is out of bounds of the mesh or the margins

Okay, I’ll investigate tomorrow. This narrows it down. It seems like maybe the Hilbert implementation is acting a bit differently as regards to out of bound points. Probably an easy fix now that I know what to look for.

thinkyhead commented:
> I can’t think of a reason why having both available at runtime is an advantage though.

I thought it might be useful for development and testing, debugging of this `P10` (Y shift? Y expansion?), without needing to do a rebuild.

marciot commented:
> > I can’t think of a reason why having both available at runtime is an advantage though.
> 
> I thought it might be useful for development and testing, debugging of this `P10` (Y shift? Y expansion?), without needing to do a rebuild.

I’ll consider that. I’ll investigate more tomorrow.

thinkyhead committed:
use REMEMBER

thinkyhead committed:
Limit to 0...N-1

thinkyhead commented:
My last commit applies a 0...N-1 limit where it was previously 0...N. That might be enough to fix this issue.

thinkyhead commented:
Ok, I think Hilbert `G26` is normal and stable now! Sorry to have hijacked this for a mini-refactor. I'm packing up all the `G29` variables into `struct`s also, and probably everything global or otherwise sharing data among groups of functions will be bound up in curly braces before long.

GhostlyCrowd commented:
> My last commit applies a 0...N-1 limit where it was previously 0...N. That might be enough to fix this issue.
> 
> `hilbert_curve::search_from_closest`
> `ExtUI::setMeshPoint`

So far I'm still getting the "out of bounds" mesh printing on Y max which is causing some silly travel. Its almost like its rolling over movements from Y max to y min because it goes from Y max Phantom mesh all the way across to Y min mesh point, and later in the mesh it goes from the Y min position and travels all the way across to print a phantom mesh point on Y max.  

This does however fix `G26 PXX`, so that it behaves, and doesn't ONLY function if `G28` happens directly before `G26 P10`

In the video, I preheat, Home then Move to X0 Y0 then do G26 P1, it moves from X0 Y0 to X 315 Y315 (which is weird it should start the mesh at X0 Y0) and starts the mesh and then you can see what happens next with the out of bounds mesh printing. 

Video quality is only SD until youtube catches up it should be HD shortly. 
https://www.youtube.com/watch?v=7XnsrsFg3kw

thisiskeithb commented:
> It would probably be nicer if the logic raised Z to 10MM after its done.

Or the defined `Z_AFTER_PROBING` point.

marciot commented:
> > My last commit applies a 0...N-1 limit where it was previously 0...N. That might be enough to fix this issue.
> > `hilbert_curve::search_from_closest`
> > `ExtUI::setMeshPoint`
> 
> So far I'm still getting the "out of bounds" mesh printing on Y max which is causing some silly travel. Its almost like its rolling over movements from Y max to y min because it goes from Y max Phantom mesh all the way across to Y min mesh point, and later in the mesh it goes from the Y min position and travels all the way across to print a phantom mesh point on Y max.
> 
> This does however fix `G26 PXX`, so that it behaves, and doesn't ONLY function if `G28` happens directly before `G26 P10`
> 
> In the video, I preheat, Home then Move to X0 Y0 then do G26 P1, it moves from X0 Y0 to X 315 Y315 (which is weird it should start the mesh at X0 Y0) and starts the mesh and then you can see what happens next with the out of bounds mesh printing.
> 
> Lastly at the end of the mesh the nozzle returns to Z0 and then drags its self across the whole mesh back to X0 Y0 instead of stopping where it finished like the old logic. It would probably be nicer if the logic raised Z to 10MM after its done.
> 
> Video quality is only SD until youtube catches up it should be HD shortly.
> https://www.youtube.com/watch?v=7XnsrsFg3kw

Can you upload your Configs? I think it likely has to do with your mesh size or bed size. Something must be overflowing and causing the wrap around you are seeing.

GhostlyCrowd commented:
[Configs_and_platformio.zip](https://github.com/MarlinFirmware/Marlin/files/6233176/Configs_and_platformio.zip)
Configs and platformio.ini, BTT SKR Pro 1v1

thinkyhead commented:
> Or the defined Z_AFTER_PROBING point.

Or the park position, if that is enabled.

GhostlyCrowd commented:
> > Or the defined Z_AFTER_PROBING point.
> 
> Or the park position, if that is enabled.

agree with this, excellent logic.

marciot commented:
@GhostlyCrowd: Can you change the line:

`return x < GRID_MAX_POINTS_X && y < GRID_MAX_POINTS_Y ? func(x, y, data) : false;`

To

`return x >= 0 && y >= 0 && x < GRID_MAX_POINTS_X && y < GRID_MAX_POINTS_Y ? func(x, y, data) : false;`

In "hilbert_curve.cpp"? x and y should never be negative, but if they are, maybe you would get the wrap around effect you are seeing. If this helps, then we may have to figure out why the values are going negative.

GhostlyCrowd commented:
> @GhostlyCrowd: Can you change the line:
> 
> `return x < GRID_MAX_POINTS_X && y < GRID_MAX_POINTS_Y ? func(x, y, data) : false;`
> 
> To
> 
> `return x >= 0 && y >= 0 && x < GRID_MAX_POINTS_X && y < GRID_MAX_POINTS_Y ? func(x, y, data) : false;`
> 
> In "hilbert_curve.cpp"? x and y should never be negative, but if they are, maybe you would get the wrap around effect you are seeing. If this helps, then we may have to figure out why the values are going negative.

can do tonight after work

marciot commented:
I was able to reproduce the problem and rule out a problem in the Hilbert curve code. The indices are coming back valid. However, the ordering of the points seem to have uncovered a bug in the G26 code.

There is a lot of strange things going on. For example, when you tell it to start at (0,0), it actually starts on the top row. This shouldn't be the case.

marciot committed:
Fix incorrect behavior on negative starting coordinates.

marciot committed:
Fix out-of-bounds error in code

marciot committed:
Fix compilation error due to function mismatch.

marciot commented:
@GhostlyCrowd: I believe I have updated this PR with fixes for all the issues. It should start in the lower-left and not do any stray moves. Please give it a whirl when you have a chance.

GhostlyCrowd commented:
> @GhostlyCrowd: I believe I have updated this PR with fixes for all the issues. It should start in the lower-left and not do any stray moves. Please give it a whirl when you have a chance.

Ok, so, The Mesh starts at X0 Y0 where my nizzle is before doing `G26 P1` the movement and the "extra" mesh is fixed. 

Two issues remain, one outstanding, when the mesh finishes the nozzle just drags its self across the whole mesh tearing it off the bed when it return to its start point X0 Y0 so reading the mesh becomes rather unusable to diagnose bed mesh issues. 

Second issue is a new one, now X max is missing the last part of the mesh as show in picture. 

![image](https://user-images.githubusercontent.com/1682110/113217810-726f5200-924c-11eb-8c03-887bdb18743c.png)

marciot committed:
Fix right hand side of mesh.

marciot commented:
@GhostlyCrowd: Okay, so I pushed a fix that should close the right hand side of the mesh while preserving the existing logic.

That said, I believe this code is highly inefficient in performance *and* memory. There is absolutely no need to scan the whole matrix for connections after each point is placed. Each new point can connect to up to four neighbors and the only information that is needed to make that determination is circle_flags, both horizontal_mesh_line_flags and vertical_mesh_line_flags are superfluous. This code can be simplified a lot.

Anyhow, this is a bold claim. I'll try to make it work now and if it does, I'll push it as a new commit :)

GhostlyCrowd commented:
> @GhostlyCrowd: Okay, so I pushed a fix that should close the right hand side of the mesh while preserving the existing logic.
> 
> That said, I believe this code is highly inefficient in performance _and_ memory. There is absolutely no need to scan the whole matrix for connections after each point is placed. Each new point can connect to up to four neighbors and the only information that is needed to make that determination is circle_flags, both horizontal_mesh_line_flags and vertical_mesh_line_flags are superfluous. This code can be simplified a lot.
> 
> Anyhow, this is a bold claim. I'll try to make it work now and if it does, I'll push it as a new commit :)

It is working almost as expected now.  the last issue is when it returns to its original position after the mesh is done the nozzle is dragging the bed.

marciot committed:
Simplify mesh connection algorithm

- Rather than checking entire mesh after each point, only check neighbors
- Eliminate looping through entire mesh
- Eliminate flags for horizontal and vertical lines

marciot commented:
@GhostlyCrowd: I implemented the new mesh connection algorithm. Way more efficient, but should look exactly the same.

> It is working almost as expected now. the last issue is when it returns to its original position after the mesh is done the nozzle is dragging the bed.

Well, that's new functionality. I'll let someone else figure out how to do that. I feel like I've done quite enough of work already that had nothing to do with my original PR.

marciot commented:
@thisiskeithb: This was one of those PRs that opened up a Pandora's box of issues and bugs. It was *supposed* to be a quick-and-easy one... 😆

marciot commented:
@GhostlyCrowd: If you don't mind retesting my latest commit, I would appreciate it. Please make sure the connections to the circles look the same as before. I changed how those were being computed and I want to make sure I didn't break the pattern.

GhostlyCrowd commented:
> @GhostlyCrowd: I implemented the new mesh connection algorithm. Way more efficient, but should behave exactly the same as the old one.
> 
> > It is working almost as expected now. the last issue is when it returns to its original position after the mesh is done the nozzle is dragging the bed.
> 
> Well, that's new functionality. I'll let someone else figure out how to do that. I feel like I've done quite enough of work already that had nothing to do with my original PR.
Its not underappreciated, do not get me wrong! 

 It's Just it is not behaving as before.  When the mesh ended the nozzle would just stay where the mesh ended, your new code drags it back to X0 Y0 and it tears the mesh off the bed.  It either is returning it to where the start position was or its doing that in design of your new code. Either way the issue is when this happens it tears the mesh off the bed where the nozzle drags which is defeating the whole purpose of G26,  If the mesh cannot be inspected for accuracy because its damaged its not worth doing at all...
 
 I hope i have not offended you, but before this PR is merged this should be addressed.  Either at end of mesh the nozzle just stops in position where it ended as was the old G26 behavior or we have to get it off the bed before it moves.

marciot commented:
@GhostlyCrowd: I believe the old code had the same bug, I just think it probably ended at a different corner of he bed and it didn't cause an issue. This was similar to the end points issue, which didn't show up until the order of the probe points got changed.

Anyhow, I don't mean to brush off your concern. It's just that all the changes I did were massaging existing logic so it did what it was supposed to be doing in the first place. New functionality is a bit harder. You have to know what things to call. I believe adding a raise nozzle would be easy for someone who knows what to call. I'm looking at you, @thinkyhead.

GhostlyCrowd commented:
> @GhostlyCrowd: I believe the old code had the same bug, I just think it probably ended at a different corner of he bed and it didn't cause an issue. This was similar to the end points issue, which didn't show up until the order of the probe points got changed.
> 
> Anyhow, I don't mean to brush off your concern. It's just that all the changes I did were massaging existing logic so it did what it was supposed to be doing in the first place. New functionality is a bit harder. You have to know what things to call. I believe adding a raise nozzle would be easy for someone who knows what to call. I'm looking at you, @thinkyhead.

Understood, thanks for all the work so far, its working great. I think @Roxy-3D  would be a good one to tag as well as she wrote the original G26 if I recall correctly.

marciot commented:
So, it looks like it may be attempting to raise the nozzle, this is line 860. So I guess the question is why it is now returning to the starting point before getting to that point. I guess I can look at it some more tomorrow. I think we're close.

GhostlyCrowd commented:
https://github.com/MarlinFirmware/Marlin/pull/21480/commits/c61d258f869a05d2883567ae72abecfcfe305bb0 The nozzle now ends the mesh at X min Y max area so when it travels back to X0 Y0 it no longer rips up the mesh but it still does not raise the z before move so I don't know if my experience will be every ones and mesh collision could still be an issue.  

Your changes are great its even more efficient then before.

thinkyhead commented:
> would be easy for someone who knows what to call. I'm looking at you, @thinkyhead

I always have to re-figure-out these things because my brain is old, but I'm willing to do that.

GhostlyCrowd commented:
> > would be easy for someone who knows what to call. I'm looking at you, @thinkyhead
> 
> I always have to re-figure-out these things because my brain is old, but I'm willing to do that.

The correct term is seasoned Veteran :P

thinkyhead committed:
Add parentheses and cell-oriented max

marciot commented on Marlin/src/module/settings.cpp:
```diff
@@ -1580,7 +1580,7 @@ void MarlinSettings::postprocess() {
         #if ENABLED(MESH_BED_LEVELING)
           if (!validating) mbl.z_offset = dummyf;
-          if (mesh_num_x == GRID_MAX_POINTS_X && mesh_num_y == GRID_MAX_POINTS_Y) {
+          if (mesh_num_x == (GRID_MAX_POINTS_X) && mesh_num_y == (GRID_MAX_POINTS_Y)) {
```
Why the parenthesis? Shouldn’t the paren be part of the macro?

thinkyhead replied:
Users are encouraged to edit these macros. Never trust users.

marciot replied:
Ah, okay. I suppose that makes sense.

marciot commented:
@thinkyhead: The reason the Z raise is not working is because at some point `move_to` was changed to take references:

```
void move_to(const float &rx, const float &ry, const float &z, const float &e_delta);
FORCE_INLINE void move_to(const xyz_pos_t &where, const float &de) { move_to(where.x, where.y, where.z, de); }
```

The code which lifts the nozzle is this:

```
  destination.z = Z_CLEARANCE_BETWEEN_PROBES;
  move_to(destination, 0);                                   // Raise the nozzle
```

However, inside "move_to", `destination` is over-written by `current_position`, the function then reads back the value it just over-wrote via the reference:

```
void move_to(const float &rx, const float &ry, const float &z, const float &e_delta) {
  destination = current_position;
  if (z != last_z) {
    last_z = destination.z = z;
    ...
  }
  ...
}
```

So, effectively, the `move_to` function ignores Z when called with an argument of "destination", which happens a lot in G26.

In many places in Marlin, I've see pass-by-value being replaced pass-by-reference for scalar types like int or float. I think this is a clean-up you like to do, but IMHO, this is poor practice, not only because it introduces aliasing bugs like this one, but because it doesn't really save any space on the stack -- pointers are often as long as the longest scalar type, so there are no saving to be had by using them.

I will push a commit that will restore "move_to" to using pass-by-value as this will fix the many places in the G26 code where Z is being ignored.

marciot commented:
I will also remove:

```
move_to(g26.xy_pos, 0); // Move back to the starting position
```

There is no need for G26 to return to the starting position. Doing so could introduces an unnecessary move when it is immediately followed by other motion commands.

marciot committed:
Fix Z-raise; do not return to start position

marciot committed:
Merge branch 'pr-ubl-hilbert-curve-followup-tweaks' of github.com:drunken-octopus/drunken-octopus-marlin into pr-ubl-hilbert-curve-followup-tweaks

marciot commented:
@thinkyhead: I have no idea what state this branch is in. It looks like some commits were already merged into "upstream/bugfix-2.0.x", others were not.

It looks like the following commits are missing in "upstream/bugfix-2.0.x":

- Fix compilation error due to function mismatch (b53d4c6)
- Simplify mesh connection algorithm (c61d258)
- Fix Z-raise; do not return to start position (ebd5965)

GhostlyCrowd commented:
Ill give this a test tonight. Looks like you've solved the quarks!

thinkyhead committed:
Merge remote-tracking branch 'upstream/bugfix-2.0.x' into pr/21480

thinkyhead committed:
constrain

thinkyhead commented:
> There is no need for G26 to return to the starting position.

I'm not sure why that behavior was wanted in the first place, except that `g26.xy_pos` can sometimes be an XY position supplied by `G26 X123 Y123`,  so maybe the next most common action would have been to adjust the Z height at that position.

Anyway, if the final Z raise now works after the `G26` procedure is done, then it seems good enough to me.

> Doing so could introduces an unnecessary move when it is immediately followed by other motion commands.

NOTE: Do not put `G26` into your slicer's starting G-code.

thinkyhead commented:
I don't think it's the use of references at all in the `move_to` situation — they are all `const` and so cannot be modified — but simply the modification of `destination`. Of course, movement functions don't have to go using `destination` for a temporary all the bloody time. They can use something else. But I am guessing that the original design was built to be dependent on `destination` but at some point there was a disconnection.

thinkyhead commented:
…at some point…
<img width="912" alt="image" src="https://user-images.githubusercontent.com/698003/113347236-79549e00-92fa-11eb-87bb-4e847475dfd1.png">

thinkyhead committed:
Revert move_to slightly

Regression from #15204

thinkyhead commented:
I did subtly alter the behavior of `move_to` during the huge conversion from arrays over to `xyz_pos_t` and the others in PR #15204. It used to be that the values of `destination.z` was not altered in the first half of `move_to` _unless_ there was a change in Z. So I've restored that behavior. I'm not sure how much depends on this small change, but I'll audit the calls to `move_to` and see where else that 'side effect' could matter.

thinkyhead commented:
> Why bother with pass-by-reference for scalar values?

It is a standard practice for the benefit of AVR, where a pointer is only 2 bytes, but a `float` is 4. At some point, when I find the time, they will all be replaced by a type (maybe called `small_float`) that adapts to the environment.

thinkyhead committed:
optimize for size

GhostlyCrowd commented:
All tested, this is a thumbs up for me on merge. 

Z gets off the mesh at the end no more nozzle drag, its all working as expected. 

Thanks for hanging in there and listening to me complain @marciot

thinkyhead committed:
Merge remote-tracking branch 'upstream/bugfix-2.0.x' into pr/21480
